torchvision/datasets/mnist.py

```diff
@@ -381,40 +381,56 @@ def __init__(
+
+    @property
+    def labels_file(self) -> str:
+        _, (url, _) = self.resources[self.subsets[self.what]]
+        return os.path.join(self.raw_folder, os.path.splitext(os.path.basename(url))[0])
+
+    def _check_exists(self) -> bool:
+        return all(check_integrity(file) for file in (self.images_file, self.labels_file))
+
+    def _load_data(self):
+        data = read_sn3_pascalvincent_tensor(self.images_file)
+        assert (data.dtype == torch.uint8)
+        assert (data.ndimension() == 3)
+
+        targets = read_sn3_pascalvincent_tensor(self.labels_file).long()
+        assert (targets.ndimension() == 2)
+
+        if self.what == 'test10k':
+            data = data[0:10000, :, :].clone()
+            targets = targets[0:10000, :].clone()
+        elif self.what == 'test50k':
+            data = data[10000:, :, :].clone()
+            targets = targets[10000:, :].clone()
+
+        return data, targets
+
     def download(self) -> None:
-        """Download the QMNIST data if it doesn't exist in processed_folder already.
+        """Download the QMNIST data if it doesn't exist already.
            Note that we only download what has been asked for (argument 'what').
         """
         if self._check_exists():
             return
+
         os.makedirs(self.raw_folder, exist_ok=True)
-        os.makedirs(self.processed_folder, exist_ok=True)
         split = self.resources[self.subsets[self.what]]
-        files = []
 
-        # download data files if not already there
         for url, md5 in split:
             filename = url.rpartition('/')[2]
             file_path = os.path.join(self.raw_folder, filename)
             if not os.path.isfile(file_path):
-                download_url(url, root=self.raw_folder, filename=filename, md5=md5)
-            files.append(file_path)
-
-        # process and save as torch files
-        print('Processing...')
-        data = read_sn3_pascalvincent_tensor(files[0])
-        assert(data.dtype == torch.uint8)
-        assert(data.ndimension() == 3)
-        targets = read_sn3_pascalvincent_tensor(files[1]).long()
-        assert(targets.ndimension() == 2)
-        if self.what == 'test10k':
-            data = data[0:10000, :, :].clone()
-            targets = targets[0:10000, :].clone()
-        if self.what == 'test50k':
-            data = data[10000:, :, :].clone()
-            targets = targets[10000:, :].clone()
-        with open(os.path.join(self.processed_folder, self.data_file), 'wb') as f:
-            torch.save((data, targets), f)
+                download_url(url, self.raw_folder, filename=filename, md5=md5)
+                if filename.endswith(".xz"):
+                    with lzma.open(file_path, "rb") as fh1, open(os.path.splitext(file_path)[0], "wb") as fh2:
+                        fh2.write(fh1.read())
+                else:
+                    extract_archive(file_path, os.path.splitext(file_path)[0])
```

Comment:
Wouldn't it be better to add support for `.xz` in `extract_archive` instead of patching it here?

Reply:
On a separate note, do we need to extract the file (saving it to a different location), instead of just using the unzipped file and calling `open_maybe_compressed_file` when loading?

Reply:
> Wouldn't it be better to add support for `.xz` in `extract_archive` instead of patching it here?

I don't think since the file is not an archive, but only compressed. The same hold for the other files with the `.gz` extension. You are still right in a way: we shouldn't use `extract_archive` for the other case either.

> On a separate note, do we need to extract the file (saving it to a different location), instead of just using the unzipped file and calling `open_maybe_compressed_file` when loading?

We could do that. I'm not sure how this would affect the performance and if it would contradict my starting assumption that loading from the original binary is not much slower than our custom binary. Do you want me to investigate this further?

Reply:
I've run the following on the `what="nist"` file of `QMNIST` since it is the largest:

```python
import timeit as _timeit
from torchvision.datasets.mnist import read_sn3_pascalvincent_tensor


def timeit(fn, number=3):
    total = _timeit.timeit(fn, number=number)
    print(f"{total / number:.2f} s")


name = "xnist-images-idx3-ubyte"
timeit(lambda: read_sn3_pascalvincent_tensor(name))
timeit(lambda: read_sn3_pascalvincent_tensor(f"{name}.xz"))
```

```
0.11 s
3.50 s
```

This is a significant difference. This is the time a user would experience every time the dataset is loaded. If we decide that this is acceptable, we can remove the extracting functionality all together. 

If not, I propose we add an `uncompress` method (maybe even to `datasets.utils`) instead of misusing the `extract_archive` function here.

Reply:
I think having `extract_archive` working for compressed files make sense to me. You can have a `.zip` compressed file with a single file within, would this be considered an `archive` or a compressed file?

I still think it makes sense to support compressed files in `extract_archive`.

> This is a significant difference. This is the time a user would experience every time the dataset is loaded. If we decide that this is acceptable, we can remove the extracting functionality all together

Yes, this is a significant difference. Let's unzip the file and use it instead